File: client/java/src/main/java/com/be/mesh/client/prsim/Network.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.prsim;

import com.be.mesh.client.annotate.Index;
import com.be.mesh.client.annotate.MPI;
import com.be.mesh.client.annotate.SPI;
import com.be.mesh.client.struct.*;

import java.util.List;

/**
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public interface Network {

    /**
     * Get the meth network environment fixed information.
     *
     * @return Fixed system information.
     */
    @MPI("mesh.net.environ")
    Environ getEnviron();

    /**
     * Check the mesh network is accessible.
     *
     * @param route Network route.
     * @return true is accessible.
     */
    @MPI("mesh.net.accessible")
    boolean accessible(@Index(value = 0, name = "route") Route route);

    /**
     * Refresh the routes to mesh network.
     *
     * @param routes Network routes.
     */
    @MPI("mesh.net.refresh")
    void refresh(@Index(value = 0, name = "routes") List<Route> routes);

    /**
     * getRoute the network edge route.
     *
     * @return edge routes
     */
    @MPI("mesh.net.edge")
    Route getRoute(@Index(value = 0, name = "node_id") String nodeId);

    /**
     * getRoutes the network edge routes.
     *
     * @return edge routes
     */
    @MPI("mesh.net.edges")
    List<Route> getRoutes();

    /**
     * GetNetDomain the network domains.
     *
     * @return net domains
     */
    @MPI("mesh.net.domains")
    List<Route> getDomains();

    /**
     * Put the network domains.
     *
     * @param domains network domains for dns resovler
     */
    @MPI("mesh.net.resolve")
    void putDomains(List<Route> domains);

    /**
     * Weave the network.
     */
    @MPI("mesh.net.weave")
    void weave(@Index(value = 0, name = "route") Route route);

    /**
     * Acknowledge the network.
     */
    @MPI("mesh.net.ack")
    void ack(@Index(value = 0, name = "route") Route route);

    /**
     * Disable the network.
     */
    @MPI("mesh.net.disable")
    void disable(@Index(value = 0, name = "node_id") String nodeId);

    /**
     * Enable the network.
     */
    @MPI("mesh.net.enable")
    void enable(@Index(value = 0, name = "node_id") String nodeId);

    /**
     * Index the network edges.
     */
    @MPI("mesh.net.index")
    Page<List<Route>> index(Paging index);

    /**
     * Network environment version.
     *
     * @return version.
     */
    @MPI("mesh.net.version")
    Versions version(@Index(value = 0, name = "node_id") String nodeId);

    /**
     * Network institutions.
     */
    @MPI("mesh.net.instx")
    Page<List<Institution>> instx(Paging index);

    /**
     * Network institutions.
     */
    @MPI("mesh.net.instr")
    void instr(List<Institution> institutions);

    /**
     * Network form alliance.
     */
    @MPI("mesh.net.ally")
    void ally(@Index(value = 0, name = "node_ids") List<String> nodeIds);

    /**
     * Network quit alliance.
     */
    @MPI("mesh.net.disband")
    void disband(@Index(value = 0, name = "node_ids") List<String> nodeIds);

    /**
     * Network feature assert.
     */
    @MPI("mesh.net.assert")
    boolean asserts(@Index(0) String feature, @Index(name = "node_ids", value = 1) List<String> nodeIds);

}

File: client/python/mesh/prsim/network.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#

from abc import abstractmethod, ABC
from typing import List

from mesh.kinds import Environ, Route, Versions, Paging, Page, Institution
from mesh.macro import mpi, spi


@spi("mesh")
class Network(ABC):

    @abstractmethod
    @mpi("mesh.net.environ")
    def get_environ(self) -> Environ:
        """
        Get the meth network environment fixed information.
        :return: Fixed system information.
        """
        pass

    @abstractmethod
    @mpi("mesh.net.accessible")
    def accessible(self, route: Route) -> bool:
        """
        Check the mesh network is accessible.
        :param route: Network route.
        :return: true is accessible.
        """
        pass

    @abstractmethod
    @mpi("mesh.net.refresh")
    def refresh(self, routes: List[Route]):
        """
        Refresh the routes to mesh network.
        :param routes: Network routes.
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.edge")
    def get_route(self, node_id: str) -> Route:
        """
        GetNetRoute the network edge routes.
        :param node_id: node id
        :return: edge routes
        """
        pass

    @abstractmethod
    @mpi("mesh.net.edges")
    def get_routes(self) -> List[Route]:
        """
        GetNetRoute the network edge routes.
        :return: edge routes
        """
        pass

    @abstractmethod
    @mpi("mesh.net.domains")
    def get_domains(self) -> List[Route]:
        """
        GetNetDomain the network domains.
        :return: net domains
        """
        pass

    @abstractmethod
    @mpi("mesh.net.resolve")
    def put_domains(self, domains: List[Route]):
        """
        Put the network domains.
        """
        pass

    @abstractmethod
    @mpi("mesh.net.weave")
    def weave(self, route: Route) -> None:
        """
        Weave the network.
        :param route:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.ack")
    def ack(self, route: Route) -> None:
        """
        Acknowledge the network.
        :param route:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.disable")
    def disable(self, node_id: str) -> None:
        """
        Disable the network.
        :param node_id:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.enable")
    def enable(self, node_id: str) -> None:
        """
        Enable the network.
        :param node_id:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.index")
    def index(self, index: Paging) -> Page[Route]:
        """
        Index the network edges.
        :param index:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.version")
    def version(self, node_id: str) -> Versions:
        """
        Network environment version.
        :param node_id:
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.instx")
    def instx(self, index: Paging) -> Page[Institution]:
        """
        Network institutions.
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.instr")
    def instr(self, institutions: List[Institution]) -> str:
        """
        Network institutions.
        :return:
        """
        pass

    @abstractmethod
    @mpi("mesh.net.ally")
    def ally(self, node_ids: List[str]) -> None:
        """
        Network form alliance.
        """
        pass

    @abstractmethod
    @mpi("mesh.net.disband")
    def disband(self, node_ids: List[str]) -> None:
        """
        Network quit alliance.
        """
        pass
